### ISSUE TITLE
`#[expect(redundant_lifetimes)]` doesn't work when `#[derive(Debug)]` is present

### ISSUE THREAD

**Issue by LuisFerLCC** [rust-lang/rust#150553]
Consider the following code:

```rust
#![deny(redundant_lifetimes)]

use std::fmt::Debug;

#[derive(Debug)]
pub struct RefWrapper<'a, T>
where
    'a: 'static,
    T: Debug,
{
    pub t_ref: &'a T
}
```

As expected, the compiler throws an error because the `'a` lifetime is redundant:

```
➜  rust-playground git:(master) ✗ cargo check
    Checking rust-playground v0.1.0 (/Users/luisferlcc/rust-playground)
error: unnecessary lifetime parameter `'a`
 --> src/lib.rs:6:23
  |
6 | pub struct RefWrapper<'a, T>
  |                       ^^
  |
  = note: you can use the `'static` lifetime directly, in place of `'a`
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(redundant_lifetimes)]
  |         ^^^^^^^^^^^^^^^^^^^

error: unnecessary lifetime parameter `'a`
 --> src/lib.rs:6:23
  |
6 | pub struct RefWrapper<'a, T>
  |                       ^^
  |
  = note: you can use the `'static` lifetime directly, in place of `'a`

error: could not compile `rust-playground` (lib) due to 2 previous errors
```

Adding `#[allow(redundant_lifetimes)]` to the type definition...

```rust
#![deny(redundant_lifetimes)]

use std::fmt::Debug;

#[derive(Debug)]
#[allow(redundant_lifetimes)]
pub struct RefWrapper<'a, T>
where
    'a: 'static,
    T: Debug,
{
    pub t_ref: &'a T
}
```

...predictably eliminates the error:

```
➜  rust-playground git:(master) ✗ cargo check
    Checking rust-playground v0.1.0 (/Users/luisferlcc/rust-playground)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
```

However, changing the `allow` attribute to `expect`...

```rust
#![deny(redundant_lifetimes)]

use std::fmt::Debug;

#[derive(Debug)]
#[expect(redundant_lifetimes)]
pub struct RefWrapper<'a, T>
where
    'a: 'static,
    T: Debug,
{
    pub t_ref: &'a T
}
```

...inexplicably fails to make the linter stand down:

```
➜  rust-playground git:(master) ✗ cargo check
    Checking rust-playground v0.1.0 (/Users/luisferlcc/rust-playground)
error: unnecessary lifetime parameter `'a`
 --> src/lib.rs:7:23
  |
7 | pub struct RefWrapper<'a, T>
  |                       ^^
  |
  = note: you can use the `'static` lifetime directly, in place of `'a`
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(redundant_lifetimes)]
  |         ^^^^^^^^^^^^^^^^^^^

error: could not compile `rust-playground` (lib) due to 1 previous error
```

But removing `#[derive(Debug)]` allows the `expect` attribute to take effect:

```rust
#![deny(redundant_lifetimes)]

use std::fmt::Debug;

#[expect(redundant_lifetimes)]
pub struct RefWrapper<'a, T>
where
    'a: 'static,
    T: Debug,
{
    pub t_ref: &'a T
}
```

```
➜  rust-playground git:(master) ✗ cargo check
    Checking rust-playground v0.1.0 (/Users/luisferlcc/rust-playground)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
```

### Meta

This issue occurs on the latest versions of all three channels:

```
➜  rust-playground git:(master) ✗ rustc --version --verbose
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: aarch64-apple-darwin
release: 1.92.0
LLVM version: 21.1.3
```

```
➜  rust-playground git:(master) ✗ rustc +beta --version --verbose
rustc 1.93.0-beta.5 (72b6488ba 2025-12-21)
binary: rustc
commit-hash: 72b6488ba485fd08cd96f194ffdd83ba8aa0b057
commit-date: 2025-12-21
host: aarch64-apple-darwin
release: 1.93.0-beta.5
LLVM version: 21.1.8
```

```
➜  rust-playground git:(master) ✗ rustc +nightly --version --verbose
rustc 1.94.0-nightly (0e8999942 2025-12-30)
binary: rustc
commit-hash: 0e8999942552691afc20495af6227eca8ab0af05
commit-date: 2025-12-30
host: aarch64-apple-darwin
release: 1.94.0-nightly
LLVM version: 21.1.8
```

<details><summary>Backtrace</summary>
<p>

```
➜  rust-playground git:(master) ✗ RUST_BACKTRACE=1 cargo check
    Checking rust-playground v0.1.0 (/Users/luisferlcc/rust-playground)
error: unnecessary lifetime parameter `'a`
 --> src/lib.rs:7:54
  |
7 | pub struct RefWrapper<#[expect(redundant_lifetimes)] 'a, T>
  |                                                      ^^
  |
  = note: you can use the `'static` lifetime directly, in place of `'a`
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(redundant_lifetimes)]
  |         ^^^^^^^^^^^^^^^^^^^

error: could not compile `rust-playground` (lib) due to 1 previous error
```

</p>
</details>

**Comment by Urgau** [comment]
The difference between the `#[allow]` case and `#[expect]` lies in the keeping of the attribute in the generated derive code.

One case see that `#[expect]` is not kept when expending the `derive(Debug)` macro.
```rust
#[expect(redundant_lifetimes)]
pub struct RefWrapper<'a, T> where 'a: 'static, T: Debug {
    pub t_ref: &'a T,
}
#[automatically_derived]
impl<'a, T: ::core::fmt::Debug> ::core::fmt::Debug for RefWrapper<'a, T> where
    'a: 'static, T: Debug {
    #[inline]
    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
        ::core::fmt::Formatter::debug_struct_field1_finish(f, "RefWrapper",
            "t_ref", &&self.t_ref)
    }
}
```

**Comment by Urgau** [comment]
We are missing the `expect` lint attribute when filtering the attributes to keep:
https://github.com/rust-lang/rust/blob/b49c7d784e96216e1cb709824629f5d53bc73201/compiler/rustc_builtin_macros/src/deriving/generic/mod.rs#L534-L551

**Comment by LuisFerLCC** [comment]
> One way I could see us fix this mismatch could be to downgrade the `#[expect]` to `#[allow]` in the derived code. That would avoid having unfilled expectations while also hiding any warnings (as asked in this issue).

That seems like a great solution, and actually one that would work for all other lints, not just for `redundant_lifetimes`. Hopefully we can implement it that way.

**Comment by Urgau** [comment]
I did some testing locally, and as I suspected, duplicating the attribute in the derived code makes the expectation fulfilled if either the original place or the derived code triggers it, satisfying T-lang model above.

Opened #152289 to duplicate the attribute and fix the bug.

**Comment by joshtriplett** [comment]
We talked about this in today's @rust-lang/lang meeting.

We were speculating about possible implementation paths (e.g. spans), but ultimately we're not trying to specify the exact implementation. The *semantic* we're trying to achieve is what we care about, and we'd really welcome suggestions for how to achieve that.

The semantic is that, unlike `allow` and `warn`, `expect` has the property of warning if zero lints are suppressed *by that specific expect*, so duplicating or consolidating `expect` attributes would result in a semantic difference (incorrectly generating or suppressing warnings about unfulfilled `expect`s).

Whether via IDs (and making sure `cfg` passes those through) or spans or some other means, what would be a viable method to ensure that an expect applied in this circumstance still handles suppressions as expected?

**Comment by Albab-Hasan** [comment]
answering @joshtriplett 's question above about a viable method.

why #152289 broke: cloning does keep the attr id but `cfg_eval` re-parses the item from tokens when it contains `#[cfg]`/`#[cfg_attr]` (`cfg_eval.rs:100`) and attr ids are not representable in tokens. so the derive input gets fresh ids while the residual item keeps the originals. their is one written attr but two ids.

tried making the ids survive first. remaping them right after `cfg_eval` in `derive.rs` fixes directly written `#[expect]` but not `#[cfg_attr(cond, expect(..))]` because under `capture_cfg` the materialized attr lives in the captured token stream and isnt in the AST attr list at that point. also tried feeding the configured item back as the residual item. it works but reverts the variant 1 semantics stabilized in #87220 (`attribute-after-derive.rs` and `derive-cfg-nested-tokens.rs` catch it).

what works here is re-land #152289 and add a fallback in `check_expectations`. id matching stays primary. an expectation unfulfilled by id also counts as fulfilled if another expectation with the same source location and lint index was fulfilled. copies of one written attr share its location and two distinct written attrs never can.

one note. `attr.span()` doesnt work as the key. cfg_attr expansion gives the materialized attr the inner `expect(..)` span on the AST path but the whole `cfg_attr` span on the captured token stream path. `path.span.to(args end)` is identical on both so i keyed on that. that span divergence might be worth fixing on its own.

semantics note since this goes beyond derives: any same-expansion span-identical copies of one written `#[expect]` now count as one expectation. e.g. a macro expanding `$i $i`. reads consistent with the model tmandry described above but its a behavior change in that corner so flagging it.

diff is the `sym::expect` one liner from #152289 plus ~30 lines in `expect.rs`. full `tests/ui` and clippys suite pass, including the #153036 and #152401 reproducers and both proc macro tests above unchanged.

@Urgau does this direction work for you? happy to open a PR if so.

**Comment by Urgau** [comment]
> any same-expansion span-identical copies of one written `#[expect]` now count as one expectation. e.g. a macro expanding `$i $i`

Unfortunately this is a real blocker, the language semantics needs to be precise, that means we can't under any circumstances rely on the spans for correctness. Spans are intended for diagnostics and can be manipulated or lost during macro expansion and serialization, as you rightfully mention.

**Comment by Albab-Hasan** [comment]
took another shot at this without spans anywhere. identity stays on attribute ids and HIR nodes.

the trick is to not copy the attr at all. the revert of #152289 stays. instead `shallow_lint_levels_on` recognizes an impl generated by a derive expansion, resolves the ADT through the impl self type and applies the ADT's `#[expect]` attrs to the impl node reusing the ADT attr's own `LintExpectationId::Stable { hir_id, attr_index, lint_index }`. a lint triggered in the derived code is suppressed and fulfills the original expectation because the id is literally the same one. only one expectation is ever registered per written attr no matter how many impls are derived. and since nothing travels through token streams anymore, the `cfg`/`cfg_attr` re-parsing that broke #152289 has nothing left to desync.

the span-identity merging you blocked on is gone entirely. macro duplicated attrs like the `$i $i` case keep distinct attribute ids and stay distinct expectations. verified behavior is identical to master on that shape.

scoping notes. the gate for "generated by derive" is the expansion kind (`MacroKind::Derive` hygiene data), not span positions. if a proc macro mangles that context the result is a no-op, current behavior, never a false merge. hand written impls for the same type are not covered, added a test locking that in. proc macro derives behave the same as master since external macro code isnt linted anyway. one limitation. early AST lints in derived code arent covered since the sharing happens at HIR level. thats status quo today so no regression.

diff is ~110 lines in `rustc_lint/src/levels.rs` plus tests. no expansion changes at all, so `attribute-after-derive` and `derive-cfg-nested-tokens` are untouched. full `tests/ui`, clippys suite including the #153036 reproducer shape, incremental, pretty and rustdoc-ui all pass.

@Urgau what do you think?

**Comment by Urgau** [comment]
I'm not sure I follow you, you want to duplicate the `struct` expectation in the `impl` declaration when building the lint levels, which is an interesting idea, but at the same you want to revert https://github.com/rust-lang/rust/pull/152289. 

Isn't the revert useless/counter-productive if we do your proposed modification in lint level building? won't this reintroduce the same issue?

You mentioned that macro hygiene could be partially mangled or absent, how reliable is that information? Did you test it with macro 2.0?

Btw, feel free to open a draft PR and ping me on it.

**Comment by Albab-Hasan** [comment]
nothing gets reverted. #153055 already landed so master has no attribute copying today and my change keeps it that way. the expectation is never duplicated either. the impl node reuses the ADT attrs own `LintExpectationId` so there is exactly one expectation per written attr and `cfg/cfg_attr` re-parsing has no second id to desync.

tested macro 2.0. `macro` wrapping the derive+expect item, `$i:item` passthrough, with `cfg` fields and `cfg_attr` expects, and a genuinely unfulfilled expect which warns exactly once. all behave correctly. the derive expansion is always the innermost one so the `MacroKind::Derive` check on the impl doesnt care what outer macro produced the item. added a test locking the macro 2.0 shapes in.

opening a draft PR.

**Comment by petrochenkov** [comment]
> We talked about this in today's [@rust-lang/lang](https://github.com/orgs/rust-lang/teams/lang) meeting.
> 
> We were speculating about possible implementation paths (e.g. spans), but ultimately we're not trying to specify the exact implementation. The _semantic_ we're trying to achieve is what we care about, and we'd really welcome suggestions for how to achieve that.
> 
> The semantic is that, unlike `allow` and `warn`, `expect` has the property of warning if zero lints are suppressed _by that specific expect_, so duplicating or consolidating `expect` attributes would result in a semantic difference (incorrectly generating or suppressing warnings about unfulfilled `expect`s).

What I dislike about this model is that it deviates from the token-based model for macros, and forces macros to maintain some additional invisible grouping that cannot be expressed in tokens and cannot be reproduced manually.

What is the token based model (I'll use attributes instead of derives for simplicity).
This code
```rust
#[my_attr] #[expect(my_lint)] struct S;
```
means that the macro `my_attr` receives tokens `#[expect(my_lint)] struct S;` as an input.
Then the macro can do whatever it wants with the tokens, including ignoring `expect` entirely and emitting `enum E` instead of `struct S`.
Built-in derive macros choose to copy some attributes from the input and re-emit them into the output impls, this is perfectly expressible in the token-based model.

In the "`expect` grouping" model the expansion infra must inspect the input tokens, recognize the `expect` attributes in them, and if the macro is invoked in a list context, reapply that `expect` to all nodes in the list after that list is produced by the macro (as tokens) and parsed.
And it must re-apply the `expect` even if the macro itself chooses to e.g. remove all the attributes from the input.
And the re-application cannot be done by generating tokens, it has to introduce some additional internal state that is maintained in parallel to the token-based representation.

"Some syntax to group items" (https://github.com/rust-lang/rust/issues/150553#issuecomment-3781254995) would resolve this concern.
The groups can be emitted as tokens, and the macros themselves could then choose to emit or not emit such groups.
Note that macro attributes and `expect` attributes are supported not only in item context, but in various other list contexts too, e.g. `function_call(#[my_attr] #[expect(my_lint)] input_tokens)`, so the groups would need to be supported there (also nested groups would need to be supported as well).

**Comment by petrochenkov** [comment]
I'm not sure how widespread the case "I want to put `expect` on a set of derived impls" is.
If it's not too widespread, then a recommendation "put the (generated) impls into a module, and put `expect` on that module" would work.
That's what is already recommended if you want to apply lint attributes to something produced by fn-like macros instead of derives.
```rust
#[expect(my_lint)]
mod impls {
  generate_impls!();
}
```

I wouldn't want to land fixes like https://github.com/rust-lang/rust/pull/158943 unless we can say that we'll definitely have groups in some time (because if we have groups, then built-in derives will use them, and the interim fix can apply only to built-in derives too).

**Comment by Albab-Hasan** [comment]
@petrochenkov the token concern doesn't apply to the pr. `#[expect]` is never copied into derive output, the revert of #152289 stays. `expn_that_defined` is set by the resolver when the output defs are collected into the AST, the spans a macro puts on its tokens have zero influence on it. a proc macro can discard every attribute it wants, it still can't affect a compiler-internal def-collector record. proc macro derives are excluded by the `is_builtin_derived` gate entirely.

the semantic above was settled by T-lang. whether to hold a fix until groups exist is their call.

on groups as a prerequisite: they have no RFC or timeline. and when they do land, builtin derives would switch to using them and this code goes away cleanly. its not a permanent commitment to the interim approach.

you said "i like the idea with `expn_that_defined`" in #158943. what specifically changed between then and now?

**Comment by petrochenkov** [comment]
>he token concern doesn't apply to the pr. #[expect] is never copied into derive output

That's exactly my point - the solution is not expressible in tokens and requires keeping some parallel non token-based information about macro expansions.

>it deviates from the token-based model for macros, and forces macros to maintain some additional invisible grouping that cannot be expressed in tokens and cannot be reproduced manually.

^^^ `expn_that_defined` is such "additional invisible grouping".
(*Some* non token-based info about macro expansion does exist today to support def-site hygiene and time-travel restrictions.)
Except it is not enough to support the grouping behavior of `#[expect]` in general case, only in those where the macro-produced nodes happen to have `DefId`s.

> you said "i like the idea with `expn_that_defined`" in [#158943](https://github.com/rust-lang/rust/pull/158943). what specifically changed between then and now?

It is better than the previous iteration of the PR, if we land some partial/hacky fix, then it will be based on `expn_that_defined`.
I still find the "wontfix" alternative (https://github.com/rust-lang/rust/issues/150553#issuecomment-5024206197) better and more principled.

**Comment by Urgau** [comment]
Renominating the issue for T-lang, specifically regarding https://github.com/rust-lang/rust/issues/150553#issuecomment-5024102888 and https://github.com/rust-lang/rust/issues/150553#issuecomment-5024206197.
